### PR TITLE
Adjust equivalent min similarity HNSW exploration logic

### LINKED ISSUE #14327: TestKnnGraph.testMultiThreadedSearch random test failure
### Description

``` 
  >     java.lang.AssertionError: [doc=0 score=0.990099 shardIndex=-1, doc=3 score=0.49751243 shardIndex=-1, doc=5 score=0.21691975 shardIndex=-1, doc=6 score=0.19960079 shardIndex=-1, doc=8 score=0.17825313 shardIndex=-1] expected:<5> but was:<8>
   >         at __randomizedtesting.SeedInfo.seed([994364CEA79E8C59]:0)
   >         at org.junit.Assert.fail(Assert.java:89)
   >         at org.junit.Assert.failNotEquals(Assert.java:835)
   >         at org.junit.Assert.assertEquals(Assert.java:647)
   >         at org.apache.lucene.index.TestKnnGraph.assertResults(TestKnnGraph.java:356)
   >         at org.apache.lucene.index.TestKnnGraph.lambda$testMultiThreadedSearch$0(TestKnnGraph.java:308)
   >         at java.base/java.lang.Thread.run(Thread.java:1575)
   > 
   >     java.lang.AssertionError: [doc=0 score=0.990099 shardIndex=-1, doc=3 score=0.49751243 shardIndex=-1, doc=5 score=0.21691975 shardIndex=-1, doc=6 score=0.19960079 shardIndex=-1, doc=8 score=0.17825313 shardIndex=-1] expected:<5> but was:<8>
   >         at __randomizedtesting.SeedInfo.seed([994364CEA79E8C59]:0)
   >         at org.junit.Assert.fail(Assert.java:89)
   >         at org.junit.Assert.failNotEquals(Assert.java:835)
   >         at org.junit.Assert.assertEquals(Assert.java:647)
   >         at org.apache.lucene.index.TestKnnGraph.assertResults(TestKnnGraph.java:356)
   >         at org.apache.lucene.index.TestKnnGraph.lambda$testMultiThreadedSearch$0(TestKnnGraph.java:308)
   >         at java.base/java.lang.Thread.run(Thread.java:1575)
  2> NOTE: reproduce with: gradlew test --tests TestKnnGraph.testMultiThreadedSearch -Dtests.seed=994364CEA79E8C59 -Dtests.locale=mgo-Latn-CM -Dtests.timezone=Asia/Dacca -Dtests.asserts=true -Dtests.file.encoding=UTF-8
```

### Version and environment details

_No response_

### PR DESCRIPTION
**Pull request by benwtrent**
It was tempting to just adjust the test data, but this seems like something the searcher should do regardless...

While the previous bug fix addressed the heat-death of the universe runtime of indexing many equal vectors, it added another weird edge case that tests picked up.

We would actually not explore a single valid candidate if it was the one that actually set the minimum score for exploration. While we shouldn't explore ALL the candidates that have the same score as the minimally accepted score, we should explore at least the first one.

This closes: https://github.com/apache/lucene/issues/14327

Its directly related to:

 - https://github.com/apache/lucene/pull/14215
 - https://github.com/apache/lucene/pull/14249

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java [hunk 1]
```diff
@@ -266,11 +266,21 @@ void searchLevel(
...
     // have to be considered.
     float minAcceptedSimilarity = Math.nextUp(results.minCompetitiveSimilarity());
+    // We should allow exploring equivalent minAcceptedSimilarity values at least once
+    boolean shouldExploreMinSim = true;
     while (candidates.size() > 0 && results.earlyTerminated() == false) {
       // get the best candidate (closest or best scoring)
       float topCandidateSimilarity = candidates.topScore();
       if (topCandidateSimilarity < minAcceptedSimilarity) {
-        break;
+        // if the similarity is equivalent to the minAcceptedSimilarity, we should explore one
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java:275]: nit: format the comment a little better (auto format seems to break it strangely)

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java [hunk 2]
```diff
@@ -266,11 +266,21 @@ void searchLevel(
...
+        // candidate
+        // however, running into many duplicates can be expensive, so we should stop exploring if
+        // equivalent minimum scores are found
+        if (shouldExploreMinSim && Math.nextUp(candidates.topScore()) == minAcceptedSimilarity) {
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java:279]: if (shouldExploreMinSim && Math.nextUp(candidates.topScore()) == minAcceptedSimilarity) => 
if (shouldExploreMinSim && Math.nextUp(topCandidateSimilarity) == minAcceptedSimilarity)

**Review by mayya-sharipova — APPROVED** [review]
Thanks Ben, makes sense to do an exploration for the first candidate with the min sim
